Support REDIRECT_URI: By adding a generic local PHP request initialisation option

bernd-wechner opened the pull request:
Lighttpd's url.rewrite does not alter the REQUEST_URI server variable but instead adds a REDIRECT_URI server variable and expects the client to honour it. 

https://redmine.lighttpd.net/projects/lighttpd/repository/14/revisions/9af58a9716b120209c4011b265657b32a414dff9

This small change respect REDIRECT_URI when it is present thus empower lighty's ur.rewrite.

The issue came to light when installing the PicoCMS app, which documents URL shortening methods in Admin settings, for Apache and Nginx. Analogous rewrite works fine in lighttpd but is delivered with this nuancal difference to the PHP environment.

If there is any risk of this misfunctioning (Apache or Nginx delivering REDIRECT_URI with some other meaning) it could be isolated and enabled with a config.php setting like "redirect_uri_enabled".

Of course, open to feedback, this is just a conversation starter if need be. It's a local patch I implemented so I can shorten PicoCMS URLs under lighttpd, and to put it into context:

1) lighttpd is not officially supported.
2) REQUEST_URI is allegedly not a CGI standard, but a PHP standard
3) REDIRECT_URI is a standard nowhere I can see bar lighttpd.
4) My local patch is not upgrade-proof, I'll lose it with upgrades so want to pursue whatever path I can to see some facility to respect REDIRECT_URI implemented in the main branch that sees it both endure here across upgrades and benefit other lightttpd users.
5) While lighttpd is not officially supported I use it to run Nextcloud and am happy to field support. It's the server of choice for embedded systems, used on OpenWRT and one reason NextcloudPi possible runs so sluggishly I had to abandon it is that it implements Apache (and every time I run htop on th Pi when it's not responding, Apache is churning - if the NextCloudPi team when lighttpd NextcloudPi might work ;-)

bernd-wechner committed:
Support REDIRECT_URI

Lighttpd's url.rewrite does not alter the REQUEST_URI server variable but instead adds a REDIRECT_URI server variable and expects the client to honour it. 

https://redmine.lighttpd.net/projects/lighttpd/repository/14/revisions/9af58a9716b120209c4011b265657b32a414dff9

This small change respect REDIRECT_URI when it is present thus empower lighty's ur.rewrite.

The issue came to light when installing the PicoCMS app, which documents URL shortening methods in Admin settings, for Apache and Nginx. Analogous rewrite works fine in lighttpd but is delivered with this nuancal difference to the PHP environment.

If there is any risk of this misfunctioning (Apache or Nginx delivering REDIRECT_URI with some other meaning) it could be isolated and enabled with a config.php setting like "redirect_uri_enabled".

kesselb commented:
Thanks for sending a pull request 👍

cc @MichaIng as the lighttpd expert :)

I remember another issue caused by lighttpd because it behaves different then apache2 or nginx. 

As we don't support lighttpd I'm against this patch. If apache2 is to heavy migrate to nginx. But both webservers performs well even on low end boxes.

bernd-wechner commented:
But why would such a tiny inconsequential change not be worthy of supporting I wonder? It enables a wonderfully light webserver. I have a lot of lighttpd infrastructure and aim for consistency, not diversification and learning the nuances of different servers.

MichaIng commented:
So this is the reason, I guess, why pretty Nextcloud URLs with Lighttpd never really worked. Of course the rewrites work but every navigation button/link brings you back to `/index.php/` URLs.

Probably more reasonable would be to open a request at Lighttpd to adapt the Apache2 and Nginx behaviour since this is hence the de-facto standard then. If Lighttpd ever want to reach a higher market share, being most compatible with Apache2 and Nginx is the key, IMO, so it should be in their interest.

We'd need to assure that `REDIRECT_URI` really matches what is expected as `REQUEST_URI` before blindly overwriting it. I found: https://httpd.apache.org/docs/2.4/custom-error.html
So `REDIRECT_URL` IS used by Apache2, at least when doing an error redirect and it does not contain the query string then but `REDIRECT_QUERY_STRING` does instead. I could not find a hint that it is set outside of error redirects, but I think it must be ruled out that this does not break anything and I don't know how to do that, e.g. checking if the webserver is Lighttpd?

bernd-wechner commented:
You are confusing Apache's `REDIRECT_URL` with `REDIRECT_URI` as only the former is mentioned on that Apache page. 

But the risk is real granted and I did say we could wrap this in a config setting.

Lightty has it's own ruling roost that doesn't want to change this, they changed it and then back again because of overarching compatibility concerns with other CGIrelated modules and that claim that neither `REQUEST_URI` nor `REDIRECT_URI` are standards at all and so should not be relied on globally. 

The PicoCMS app however respects rewrites configured per its shortURL format `/sites/` and doesn't direct back to `index.php` and that is both the context in which this need was discovered and arose.

But there's no need to check if the server is this or that if a simple config setting could enable and have disabled by default, respect of such a redirection. I mean I do agree that it's as much a lighty issue as a Nextcloud issue as an Apache issue as an Nginx issue, and that it relates to PHP's documentation of allegedly (I'm not pro here) non a standard CGI server variable, REQUEST_URI, encouraging PHP devs to rely on it.

Lighty sent me here, and Id id warn them that opens its whole own can of worms with another dev team and different vies and so on ;-). 

Anyhow, wrapping it in a config setting seems innocuous and at worst could be criticsed as a bell/whistle as it's not even bloating, it's a teeny little tweak to keep conflicting dev community standards workable together.

I mean there's a real daft superbly awkward alternative, in that I write a patcher that puts that snippet in index.PHP as the first lines after any upgrade ;-). But I much prefer negotiated solutions, team work, community and inter community interoperability if at all possible. If that makes any sense.

kesselb commented:
> Anyhow, wrapping it in a config setting seems innocuous and at worst could be criticsed as a bell/whistle as it's not even bloating, it's a teeny little tweak to keep conflicting dev community standards workable together.

And we have to support it forever. For a webserver with no relevance.

MichaIng commented:
Ah indeed, sorry for mixing up `REDIRECT_URI` and `REDIRECT_URL` 😄.

At least `REQUEST_URI` is documented in [PHP docs](https://www.php.net/manual/en/reserved.variables.server.php) and [Apache2 docs](https://httpd.apache.org/docs/2.4/expr.html) and [Nginx docs](https://nginx.org/en/docs/http/ngx_http_core_module.html#var_request_uri) (if this is what is stored in PHP `$_SERVER` array with capital letters?), so regardless if its "official" standard, it's a de-facto standard that is used by web developers, while `REDIRECT_URI` is obviously only rarely when someone is keen to do the additional effort for Lighttpd only.

Regardless, is not the only case where Lighttpd made decisions which intentionally or accidentally broke PHP scripts (Nextcloud as well in another context), the question is if this very simple and clear workaround is safe to apply (i.e. assured to not break anything else) or not. I also like Lighttpd for small low traffic low resource home servers, but I understand that there is a limit about adding compatibility code.

bernd-wechner committed:
A more generic solution to supporting REQUEST rewrites

This is a generic implementation that does not add any config variables per se (though it could rest on one to name the init file) but requires one server administrator written configuration that can run some PHP code early in the request cycle. This can open the door to supporting any number of other weird and wonderful web servers perhaps that don't even popular REQUEST_URI.

bernd-wechner commented:
> And we have to support it forever. For a webserver with no relevance.

I'm not a great fan of unnecessary disparagement to be honest and a much greater fan of collaboration and mutual support.
Popularity and relevence are not one and the same, and it's of course no big issue if popularity is the guiding principle, if stated as thus, rather than relevance. And yes nginx is a fine server that performs on par with (better in some ways) that lighty, but lighty remains the choice implemented on OpenWRT. works fine and powers some interesting sites:

https://redmine.lighttpd.net/projects/lighttpd/wiki/PoweredByLighttpd

and has its fans:

"So like most every other PHP app I've worked with, everybody seems to support both apache and nginx but poor old lighttpd gets left out in the cold and has no support on how to configure it in documentation.

Well have no fear, because you'll have to pry lighttpd out of my cold dead hands!" - jsebean
https://jsebean.com/2017/01/happy-monday-getting-yourls-to-work-with-lighttpd-support

On my side I will perhaps one day branch out to nginx (because it does perform on par with lighty and is more popular) but in the interim I run several servers that use lighty and lighty alone simply because I branched out from OpenWRT and didn't feel like learning the idiosynchrascies of other servers with divers maintenance needs).

Apache remains the scourge of the FOSS server world which like your comment is disparaging but in this case not merely a value assessment or popularity assessment but based on a goodly number of perforance trials between it and its usurper nginx, and that I see it used in resource poor ARM based architectures like NextCloudPi and the QNAP OS (which blows my mind when I look at either one on its knees and get enough response to run an htop and see the whole box is brought to its knees by ... apache). 

So, that digression aside, and understanding the maintenance concerns and risks of the proposed patch (which I stated up front was to drive conversation that I always harbour the hope will be collaborative solution seeking rather than dismissive).

My take home from your comment is that a config setting is seen as a high maintenance cost, at least too high to warrant adding one for a small community of users. But it as ever opens some questions:

1. Where is the cost in (the maintenance of) a config setting? Not least for an undocumented one say?

2. Is there any (other) means of injecting some bootstrap code that is run whenever a page is loaded (index.php is run)? If not in the current core code, can an app achieve that? I notice for example that appinfo/app.php on every enabled app is required and sees $_SERVER. Moreover as this requirement is fairly specific to the PicoCMS app, so I moved that snippet into its app.info and it doesn't work, it runs, but too late and REQUEST_URI has been used already by Nextcloud somewhen prior.  

If the answer to 2. is No. Then:

3. Can we make a case that this should be supported. That beside config.php we could support another file, be it `init.php` or `request.php` but either way it is a more generic form of config that would permit local request handling intervention configurable by a server admin much like config.php and php.ini are. I pushed a sample that works fine to this PR.

But as stated, I am a) very grateful for your time and energy, and b) request similar respect for the time and energy I am investing and c) seek to negotiate a workable solution. And yes I've plied the lighty end:

https://redmine.lighttpd.net/boards/2/topics/9316

I am piggy in the middle, and seeking amicable, negotiators not curt dismissals no requests one side or the other to abandon that application in favour of  another.

bernd-wechner commented:
I will argue that:

1. This approach is very empowerig for Nexcloud admins.
2. It does permit code injection via the required file, but poses not greater risk than any other php file in the Nexcloud code base, which can be modified by anyone which an modify these config files.
3. By living in the config directory it should (I presume) be upgrade proof, that is survive Nextcloud upgrades in the same way config.php does. There being a slight chance that there is a code dependency I haven't spotted in that (that the upgrade code specifically exempts confi.php not the config directory and that this init file would require similar exemption in that case, somewhere in the code base.

bernd-wechner commented:
> Ah indeed, sorry for mixing up `REDIRECT_URI` and `REDIRECT_URL`

Yes, REQUEST_URI is documented in PHP and I do wish lighty would respect it, and concur it's in everyone's interest to converge on de fact standards where formal standards are loose too. But alas no luck at lighty. And there's a chance that lighty with its mod_magnet can do the trick, but it doesn't look easier than a tiny PHP tweak. And I admit I like my last one ... ;-). I'm keen to hear counter arguments (always open to learning and constructive dialog!).

MichaIng commented:
One minor security impact: An attacker only requires write access to the config directory now instead of write access to an actual file.
But I think this can be ignored, also the creation of an additional `.htaccess` file in a subdir is a similar risk and usually directory and file write permissions/umask is the same.
And this indeed might enable admins to do a bunch of other stuff they currently need to edit the Nextcloud PHP scripts for, breaking code integrity and re-applying after every update.

Ah, talking about code integrity, this file needs to be added to the exclusion list for the code integrity check then, otherwise admin panel and updater will report an "additional file" error.

bernd-wechner commented:
> Ah, talking about code integrity, this file needs to be added to the exclusion list for the code integrity check then, otherwise admin panel and updater will report an "additional file" error.

Thanks, I love creative thinking and have worked for years in dev, documentation, and QA across the field. Where can I find this exclusion list? Any tip appreciated. But I will of course just search for "exclusion" for starters and maybe "integrity" based on your clues.

As an aside reading the PHP docs they write:

>  'REQUEST_URI'
>     The URI which was given in order to access this page; for instance, '/index.html'. 

And Apache's docs write:

> REQUEST_URI
>     The path component of the requested URI, such as "/index.html". This notably excludes the query string which is available as its own variable named QUERY_STRING.

And that does sort of defend lighty's position some. As even after a rewrite this holds true, as in I did access the site with /sites/.  But I guess it rests on interpretation, given or requested by the user or given by the web server. Just a curio.

By the by, I actually believe this idea of an init.php really provide power to admins in diverse environments with diverse needs. This is but one application.

bernd-wechner commented:
OK had a poke around and I think you meant this list:

		$excludedElements = [
			'.well-known',
			'data',
			'index.php',
			'status.php',
			'remote.php',
			'public.php',
			'ocs/v1.php',
			'ocs/v2.php',
			'config',
			'themes',
			'apps',
			'updater',
		];

which appears in `Updater::deleteOldFiles`. But if I read tht right it excludes the whole config directory so we should be right. It does take explicit steps (oddly) to remove the sampel config:

		$configSampleFile = $this->baseDir . '/../config/config.sample.php';
		if(file_exists($configSampleFile)) {
			$this->silentLog('[info] config sample exists');

			// Delete example config
			$state = unlink($configSampleFile);
			if ($state === false) {
				throw new \Exception('Could not unlink sample config');
			}
		}

but nothing else in the config directory. So an optional init.php living there would (it seems) survive an upgrade.

Is there somewhere else that has an exclusion list integrity check?

MichaIng commented:
That seems to be the list that of files to leave in place before merging the new files during update. `config.sample.php` is likely removed since a new sample config one is shipped with the update.

I meant the integrity check which excludes e.g. a few known auto-generated files/dirs or such which are expected to be changed, like `.well-known`, `.rnd`, `config/config.php`, `.user.ini` and few others. Any file which is optional but expected or expected to be changed by the admin, must be treated specially. Let me see if I can find the part of the code.

bernd-wechner commented:
So, we're agreed, if it's in the config directory, it's safe from the integrity check and being clobbered by an upgrade?

MichaIng commented:
Jep seems to be from what we found, but a quick test is always great. I'll apply NC20 beta soon, will add the file and see what happens.

kesselb commented:
> By the by, I actually believe this idea of an init.php really provide power to admins in diverse environments with diverse needs.

It's a nightmare to debug issues if admin's are able to hook / manipulate the bootstrapping process at such an early stage.

> That beside config.php we could support another file, be it init.php or request.php but either way it is a more generic form of config that would permit local request handling intervention configurable by a server admin much like config.php and php.ini are.

https://www.php.net/manual/en/ini.core.php#ini.auto-prepend-file should work for your REQUEST_URI = REDIRECT_URI case. 

> https://redmine.lighttpd.net/projects/lighttpd/wiki/PoweredByLighttpd

```
curl -I https://api.alitools.io/
server: nginx
```

```
curl -I https://domainparking.ru/
server: nginx
```

```
curl -I https://www.gesichterparty.de/
Server: Apache
```

```
curl -I https://olhares.com/
Server: nginx/1.8.1
```

```
https://www.phpbb8.de/
Also nginx for hosted forums
```

```
curl -I https://www.pricearchive.org/
server: nginx
```
```

curl -I http://zone.webgarden.cz/
Server: nginx/1.6.2
```

:man_shrugging: 

> seek to negotiate a workable solution.

I've used lighttpd some years back. It's a stable and reliable web server. For some reason people pick Nginx or Apache2 today. I don't have a strong opinion on that. For most websites the web server is not a mission critical decision.

If we add a feature here it's nearly impossible to remove it again. Let's assume we add a undocumented way to run Nextcloud on lighttpd with clean urls. People start using it, building their setups around it and a few versions later something breaks. It breaks because no one is testing Nextcloud on lighttpd but people demand us to fix it and are disappointed because we don't fix it.

I don't see a benefit for Nextcloud by adding specific code to support clean urls on lighttpd. It's enough work already (on GitHub and https://help.nextcloud.com/) to support Apache2 and Nginx.

bernd-wechner commented:
@kesselb, Aside from an initially dismissive judgmental tone, you are indeed a a PHP genius. Just tried the `auto-prepend-file` feature and it works beautifully. Thus, this PR has served it purpose from my end (stimulating conversation to solve the problem). But I do, with due respect, find your fears a little exaggerated.  For a number of reasons (by the by, but in the spirit of constructive conversation I offer it as feedback not meaning to be critical):

1. The ability to inject admin provided code can't be quite that scare given PHP already supports it ;-). 
2. It's also pretty common elsewhere, I know Djnago for instance supports arbitrary middleware and I wrote a little piece of middleware for it to solve a similar problem (not the same one but similarly related to a need for early intervention between the server and the Django code base).
3. To support something generic like a middleware injector (this `ini.php` idea) is not specific to this problem (hence the beautiy of it) but of generic use and empowering, and not something that would break. It's a great place to stick `phpinfo();` for example perhaps conditionally called for debugging locally without having to touch index.php (which would work on something like the snap build if the config dir is writeable but the rest isn't - I abandoned the snap build because of the impossibility of debugging it when something went awry, as it did when I turned on SSL - long since diagnosed and fixed by going an ordinary Nextcloud install - albeit behind a server I'm familiar with maintaining and consistent with the environment here).
4. Your concerns about people using a feature (to support lighty say) and then being upset if they don't get support seems to contradict the earlier (and right) observation that the lighty community is by comparison very small, and I would add though, due probably to that same fact, being much less popular, very accustomed to solving problems the hard way and dealing with "being left out". For which reason mod_magnet exists in lighty (that's the light side of middleware injection), but it's a little clunky, a real option, but not  as nice as a few lines of PHP.

Still, it's all moot, and I thank you for your time, and input as `auto-prepend-file` is indeed the code injection point that I needed that is Nextcloud upgrade proof, and doesn't involve over complex manipulations.

I still believe ti a generally good idea for Nextcloud (or any similar software) to support admin generated code injection early in the request handling cycle, but I won't push that barrow as I've plenty else to do. Django BTW handles it very elegantly in a way that allows code wrapping around the request (so you can run code before and after it).

As an aside, I'll let Lighty folk know their page on sites powered by it is dated ;-). It does seem folk go to nginx. Who knows why historically. It was clear that Apache had seen its day at some point (both lighty and nginx outperform it significantly) and once can only speculate.

Thanks again! I'm rolling.

MichaIng commented:
Great thing with `auto_prepend_file`, I guess this can be even used in `.user.ini`, right? There is even [auto_append_file](https://www.php.net/manual/en/ini.core.php#ini.auto-append-file) which then together covers what Django implemented?
______________
Further thoughts if you have some boring hours:
 
I understand the argument to not officially implement/support such things that potentially cause hard to debug issues, although moving a single file out of place should be assumed to be a simple task to rule it out. What PHP itself allows is a different question, also the webserver or the database server can be badly configured to cause any thinkable issue, but when Nextcloud offer an additional feature itself surely it is expected to work safe, reliable, long-term etc and it is always a different thing when an issue is caused by something that Nextcloud offers itself compared to something with the surrounding environment that is out of Nextcloud's control. With a bid bad luck, one could find 3rd party guides after a while that suggest to place such an inclusion script for this or that reason and inexperienced users might follow those via copy&paste without thoroughly understanding it or even forgetting about it and later land with related issues on: https://help.nextcloud.com/

> "With Great Power Comes Great Responsibility"

😄

But on the other hand, https://help.nextcloud.com/ is a community support forum, not something where anyone can "expect" support and a solution or where anyone is forced to invest time to help. Of course many people, including myself by times, investing much time there to help admin/user but with or without such a script inclusion/hook feature there are cases where no-one is able to help or the issue is finally not related to Nextcloud at all but the surrounding environment, so I don't see such really increasing the community support effort. For paid support it is a different question, but checking a single file should be doable. Regardless, for now this discussion doesn't need to be done anymore 😃.

The other Lighttpd-related issue is btw: #17065
Lighttpd does some URL (including query string) decoding (partly) and normalisation to have a consistent and supposedly more secure URL. The issue is/was that `%2B` in the query string was decoded to `+` and PHP decoded `+` to white space, which breaks the Nextcloud access token that can contain `+` signs and by this most importantly breaks logout (a problem on public PCs). The `%2B` decoding has been removed in Lighttpd now, but the "broken" version is currently shipped in Debian Buster and likely will not be updated, at least not outside of backports. `lighttpd.conf` can be configured to disable this URL normalisation but it's enabled by default, means that Nextcloud logout is by default broken with Lighttpd on current Debian stable release...
I had some discussion with the Lighttpd dev that implemented this URL normalisation and I'm still not convinced that such is a great idea at all, especially when the query string is included, as I don't see a way to reliable know which "code" is really meant as URL encode and which is just a random string that by chance matches an URL-code and if any handler behind the webserver (like PHP) already does URL decoding or not. IM(unqualified)O any potential security risk should be handled webserver-internally (e.g. when its about access permissions) without manipulating the URL that is forwarded to another handler. And at least during my tests Lighttpd successfully prevents any unauthorised access, like using `..` in URL to escape blocked access and such, without having the additional URL options enabled, just as one would expect from a webserver for sure. But actually I lack insights into webserver development to conduct such discussions 😉. From my sight it just looks like that Lighttpd decision policy is not always 100% nice to web application developers, leading to missing or limited support in cases and such I guess can lead to decreasing marked share as well. And this is sad to see, as I like Lighttpd as well for many use cases and I generally like concurrence among products I use 😉.
